File: apps/oauth_demo/oauth.py
import os
from urllib.parse import quote

import httpx

ROUTER_URL = os.environ.get("OPENHOST_ROUTER_URL")
APP_TOKEN = os.environ.get("OPENHOST_APP_TOKEN")
APP_NAME = os.environ.get("OPENHOST_APP_NAME")
ZONE_DOMAIN = os.environ.get("OPENHOST_ZONE_DOMAIN")

OAUTH_SERVICE_URL = "github.com/imbue-openhost/openhost/services/oauth"
_ENCODED_SVC = quote(OAUTH_SERVICE_URL, safe="")
OAUTH_V2_BASE = f"{ROUTER_URL}/_services_v2/{_ENCODED_SVC}"

_mock_base_url: str | None = None


def set_mock_oauth_url(url: str | None) -> None:
    global _mock_base_url
    _mock_base_url = url.rstrip("/") if url else None


def _base_url() -> str:
    return _mock_base_url or OAUTH_V2_BASE


GMAIL_SCOPE = "https://www.googleapis.com/auth/gmail.readonly"
GITHUB_SCOPE = "repo"

AUTH_HEADERS = {"Authorization": f"Bearer {APP_TOKEN}"}

SCOPE_MAP = {
    "google": [GMAIL_SCOPE],
    "github": [GITHUB_SCOPE],
}

PROVIDERS = [
    {
        "name": "google",
        "label": "Google",
        "action_path": "unread",
        "action_label": "Emails",
    },
    {
        "name": "github",
        "label": "GitHub",
        "action_path": "repos",
        "action_label": "Repos",
    },
]


class AuthRedirectRequired(Exception):
    """User needs to visit a URL to grant permissions or complete OAuth consent."""

    def __init__(self, url: str):
        self.url = url


class OAuthError(Exception):
    """OAuth token request failed with a displayable error."""


class OAuthServiceUnavailable(Exception):
    """The OAuth service is not installed or not running."""


class _PermissionDenied(Exception):
    def __init__(self, approve_url: str):
        self.approve_url = approve_url


class _OAuthConsentRequired(Exception):
    def __init__(self, authorize_url: str):
        self.authorize_url = authorize_url


def _check_error_response(resp: httpx.Response) -> None:
    """Raise a typed exception for any non-success response."""
    if resp.is_success:
        return
    try:
        data = resp.json()
    except Exception as e:
        raise OAuthError(f"Request failed: {resp.status_code} {resp.text}") from e
    error = data.get("error")
    if error in ("service_not_found", "service_not_running", "service_not_available"):
        raise OAuthServiceUnavailable(data.get("message", "OAuth service unavailable"))
    # V2 permission_required (from provider via router)
    if error == "permission_required":
        grants = data.get("grants_needed", [])
        url = ""
        if grants:
            url = grants[0].get("grant_url") or grants[0].get("approve_url", "")
        raise _PermissionDenied(approve_url=url)
    # V1 permission_denied
    if error == "permission_denied":
        raise _PermissionDenied(approve_url=data.get("approve_url", ""))
    if resp.status_code == 401 and data.get("authorize_url"):
        raise _OAuthConsentRequired(data["authorize_url"])
    raise OAuthError(data.get("message", f"Request failed: {resp.status_code}"))


async def get_accounts(provider: str) -> list[str]:
    """Get list of connected account labels for a provider.

    Returns empty list if permissions haven't been granted yet.
    """
    scopes = SCOPE_MAP.get(provider)
    if not scopes:
        raise ValueError(f"Unknown provider: {provider}")
    async with httpx.AsyncClient(timeout=10) as client:
        resp = await client.post(
            f"{_base_url()}/accounts?version=>=0.1.0",
            json={"provider": provider, "scopes": scopes},
            headers=AUTH_HEADERS,
        )
        try:
            _check_error_response(resp)
        except _PermissionDenied:
            return []
        return resp.json().get("accounts", [])


async def get_oauth_token(provider: str, scopes: list[str], account: str = "default", return_to: str = "") -> str:
    """Get an OAuth access token from the secrets service.

    Args:
        return_to: URL to redirect back to after auth completes.
                   Defaults to //{APP_NAME}.{ZONE_DOMAIN}/client/oauth-complete.

    Returns the access token string.

    Raises:
        AuthRedirectRequired: User needs to visit a URL (permissions or OAuth consent).
        OAuthServiceUnavailable: The OAuth service is not installed or not running.
        OAuthError: Other failures.
    """
    if not return_to:
        return_to = f"//{APP_NAME}.{ZONE_DOMAIN}/client/oauth-complete"
    try:
        async with httpx.AsyncClient(timeout=10) as client:
            resp = await client.post(
                f"{_base_url()}/token?version=>=0.1.0",
                json={
                    "provider": provider,
                    "scopes": scopes,
                    "return_to": return_to,
                    "account": account,
                },
                headers=AUTH_HEADERS,
            )
    except httpx.HTTPError as e:
        raise OAuthError(f"Token request failed: {e}") from e

    try:
        _check_error_response(resp)
    except _PermissionDenied as e:
        raise AuthRedirectRequired(e.approve_url) from e
    except _OAuthConsentRequired as e:
        raise AuthRedirectRequired(e.authorize_url) from e

    return resp.json()["access_token"]

File: apps/oauth_demo/static/oauth.js
/**
 * Client-side OAuth helper library.
 *
 * Talks to the router's V2 service endpoints via cross-origin fetch:
 *   POST /_services_v2/<encoded-oauth-url>/accounts  — list connected accounts
 *   POST /_services_v2/<encoded-oauth-url>/token     — get access token
 *
 * Usage:
 *   const oauth = new OAuthClient({
 *     scopes: { google: ['https://...gmail...'], github: ['repo'] },
 *     appName: 'oauth-demo',
 *     zoneDomain: 'user.host.imbue.com',
 *   });
 *   const accounts = await oauth.getAccounts('google');
 *   const token = await oauth.getToken('google', 'user@gmail.com');
 *   await oauth.connect('google');  // must be called from a click handler
 */

var OAUTH_SERVICE_URL = "github.com/imbue-openhost/openhost/services/oauth";
var ENCODED_OAUTH_URL = encodeURIComponent(OAUTH_SERVICE_URL);

class OAuthClient {
  constructor({ scopes, appName, zoneDomain }) {
    this.scopes = scopes;
    this.appName = appName;
    this.zoneDomain = zoneDomain;
    this._popup = null;
  }

  _scopesFor(provider) {
    var s = this.scopes[provider];
    if (!s) throw new OAuthError("Unknown provider: " + provider);
    return s;
  }

  _returnTo() {
    return "//" + this.appName + "." + this.zoneDomain + "/client/oauth-complete";
  }

  /** POST to a V2 service endpoint with credentials. */
  _serviceFetch(endpoint, body) {
    var url = "//" + this.zoneDomain + "/_services_v2/" + ENCODED_OAUTH_URL +
      "/" + endpoint + "?version=>=0.1.0";
    return fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      credentials: "include",
      body: typeof body === "string" ? body : JSON.stringify(body),
    });
  }

  /**
   * Parse an error response. Returns:
   *   { type: "permission", url } for 403 permission_required (grant_url or approve_url)
   *   { type: "oauth", authorize_url } for 401 authorization_required
   *   { type: "error", message } for other errors
   *   null if response is ok
   */
  async _parseResponse(resp) {
    if (resp.ok) return null;
    var data;
    try { data = await resp.clone().json(); } catch (e) {
      return { type: "error", message: "Request failed: " + resp.status };
    }
    if (data.error === "permission_required" && data.grants_needed) {
      var grant = data.grants_needed[0] || {};
      var url = grant.grant_url || grant.approve_url || "";
      return { type: "permission", url: url };
    }
    if (data.approve_url) return { type: "permission", url: data.approve_url };
    if (data.authorize_url) return { type: "oauth", authorize_url: data.authorize_url };
    return { type: "error", message: data.message || "Request failed: " + resp.status };
  }

  /**
   * List connected accounts for a provider.
   * Throws OAuthPermissionError if permissions are needed.
   * Returns empty array on other errors.
   */
  async getAccounts(provider) {
    var resp = await this._serviceFetch("accounts", {
      provider: provider,
      scopes: this._scopesFor(provider),
    });
    var err = await this._parseResponse(resp);
    if (!err) return (await resp.json()).accounts;
    if (err.type === "permission") throw new OAuthPermissionError(err.url);
    return [];
  }

  /**
   * Get an OAuth access token. Must be called from a click handler.
   * Throws OAuthPermissionError if permissions need granting first.
   * Handles OAuth consent via popup automatically.
   */
  async getToken(provider, account = "default") {
    var resp = await this._serviceFetch("token", {
      provider: provider,
      scopes: this._scopesFor(provider),
      return_to: this._returnTo(),
      account: account,
    });
    var err = await this._parseResponse(resp);
    if (!err) return (await resp.json()).access_token;
    if (err.type === "permission") throw new OAuthPermissionError(err.url);
    if (err.type === "oauth") {
      await this.openPopup(err.authorize_url);
      // Retry after OAuth
      resp = await this._serviceFetch("token", {
        provider: provider,
        scopes: this._scopesFor(provider),
        return_to: this._returnTo(),
        account: account,
      });
      if (resp.ok) return (await resp.json()).access_token;
    }
    throw new OAuthError(err.message || "Token request failed");
  }

  /**
   * Connect a new account via popup. Must be called from a click handler.
   * Throws OAuthPermissionError if permissions need granting first.
   */
  async connect(provider, account = "new") {
    var resp = await this._serviceFetch("token", {
      provider: provider,
      scopes: this._scopesFor(provider),
      return_to: this._returnTo(),
      account: account,
    });
    if (resp.ok) return;
    var err = await this._parseResponse(resp);
    if (err.type === "permission") throw new OAuthPermissionError(err.url);
    if (err.type === "oauth") {
      await this.openPopup(err.authorize_url);
      return;
    }
    throw new OAuthError(err.message || "Connect failed");
  }

  /** Open a URL in a popup and wait for an auth_complete postMessage. */
  openPopup(url) {
    return new Promise(function (resolve, reject) {
      if (this._popup && !this._popup.closed) this._popup.close();
      this._popup = window.open(url, "auth_popup");

      if (!this._popup) {
        reject(new OAuthError("Popup blocked. Please allow popups and try again."));
        return;
      }

      var popup = this._popup;
      var cleanup = function () { window.removeEventListener("message", onMsg); clearInterval(poll); };
      var onMsg = function (e) {
        if (e.data && e.data.type === "auth_complete") { cleanup(); resolve(); }
      };
      var poll = setInterval(function () {
        if (popup.closed) { cleanup(); resolve(); }
      }, 500);
      window.addEventListener("message", onMsg);
    }.bind(this));
  }
}

class OAuthError extends Error {
  constructor(message) { super(message); this.name = "OAuthError"; }
}

class OAuthPermissionError extends Error {
  constructor(url) {
    super("Permissions required");
    this.name = "OAuthPermissionError";
    this.url = url;
  }
}
